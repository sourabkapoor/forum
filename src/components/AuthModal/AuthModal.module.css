@font-face {
  font-family: "Google Sans";
  src: url(../../assets/ProductSans-Regular.ttf);
}
.sign-up-modal {
  color: black;
  display: grid;
  place-items: center;
  width: 100%;
  height: 100%;
}

.sign-up-modal__container {
  background-color: white;
  box-sizing: border-box;
  border-radius: 6px;
  overflow-y: auto;
  box-shadow: 0 2px 20px 0 rgb(0, 0, 0 / 30%);
  pointer-events: auto;
  display: flex;
  align-items: center;
  width: 850px;
  height: 650px;
  position: relative;
}

.sign-up-modal__exit-icon {
  width: 24px;
  height: 24px;
  object-fit: contain;
  filter: brightness(0) invert(0);
  flex-grow: 0;
}

.sign-up-modal__oauth {
  font-family: "Google Sans";
  font-weight: 500;
  text-overflow: ellipsis;
  overflow: hidden;
  padding: 2px 10px;
  width: 200px;
  max-width: 400px;
  min-width: min-content;
  display: flex;
  align-items: center;
  margin: 0 auto;
  border-radius: 20px;
  height: 40px;
  letter-spacing: 0.25px;
  cursor: pointer;
  border: 1px solid #dadce0;
  color: black;
  width: 100%;
}
.sign-up-modal__oauth:hover {
  color: #3c4043;
}

.sign-up-modal__oauth-text {
  flex-grow: 1;
}

.sign-up-modal__divider {
  display: flex;
  align-items: center;
  margin: 28px 0;
  justify-content: space-between;
}

.sign-up-modal__divider-line {
  border-top: 1px solid #edeff1;
  width: 40%;
}

.sign-up-modal__divider-text {
  font-size: 14px;
  font-weight: 500;
  line-height: 18px;
  color: white;
}

.sign-up-modal__exit {
  position: absolute;
  background-color: inherit;
  border: none;
  right: 16px;
  top: 16px;
  cursor: pointer;
}
.sign-up-modal__icon {
  width: 18px;
  height: 18px;
}

.sign-up-modal__title {
  font-size: 18px;
  font-weight: 500;
  line-height: 22px;
}

.sign-up-modal__description {
  margin-bottom: 48px;
  font-size: 12px;
  font-weight: 400;
  line-height: 18px;
  margin-top: 8px;
  font-family: inherit;
  color: var(--secondary-color);
}

.sign-up-modal__link {
  color: #0079d3;
  text-decoration: none;
  cursor: pointer;
}

.sign-up-modal__image {
  width: 156px;
  background-color: rgb(213, 176, 250);
  /* background-image: url("https://www.redditstatic.com/accountmanager/bbb584033aa89e39bad69436c504c9bd.png"); */
  height: 100%;
  background-repeat: no-repeat;
  background-size: cover;
}

.sign-up-modal__content {
  padding: 24px;
}

.sign-up-modal__form {
  width: 280px;
}

.sign-up-modal__input {
  outline: 0;
  width: 100%;
  transition: all 0.2s ease-in-out;
  padding: 22px 12px 18px;
  height: 12px;
  box-sizing: border-box;
  border: 1px solid rgb(0, 0, 0, 0.1);
  border-radius: 4px;
  background-color: #fcfcfb;
  font-size: 14px;
  font-weight: 400;
  line-height: 21px;
  font-family: inherit;
  border: 2px solid #24a0ed;
  margin-top: 24px;
}

.sign-up-modal__button {
  font-family: inherit;
  font-size: 14px;
  font-weight: 700;
  letter-spacing: unset;
  line-height: 18px;
  text-transform: unset;
  background: #0079d3;
  border-radius: 999px;
  color: #fff;
  height: 40px;
  padding: 0 16px;
  border: none;
  width: 100%;
  margin-top: 12px;
  cursor: pointer;
}
.sign-up-modal__button:hover {
  background-color: #1484d7;
}

.sign-up-modal__bottom-text-container {
  margin-top: 24px;
  font-size: 12px;
  font-weight: 400;
  line-height: 18px;
}

.sign-up-modal__bottom-text {
  font-size: 12px;
  font-weight: 600;
  letter-spacing: 0.5px;
  line-height: 24px;
  text-transform: uppercase;
  color: #0079d3;
  text-decoration: none;
  padding-left: 8px;
  background-color: transparent;
  border: none;
  cursor: pointer;
}

@media(max-width: 930px) {
  .sign-up-modal__container {
    width: 85%;
  }
}