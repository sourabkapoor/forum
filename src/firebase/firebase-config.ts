// Import the functions you need from the SDKs you need

import { initializeApp } from "firebase/app";
import { getStorage } from "firebase/storage";

// TODO: Add SDKs for Firebase products that you want to use

// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration

const firebaseConfig = {
  // apiKey: "AIzaSyCLwXf77NuoR0ESLwAAgez_wS_gM8V4J7M",

  // authDomain: "reddit-clone-23782.firebaseapp.com",

  // projectId: "reddit-clone-23782",

  // storageBucket: "reddit-clone-23782.appspot.com",

  // messagingSenderId: "855149209259",

  // appId: "1:855149209259:web:cb03f2ae83010e6feab9ee",

  // measurementId: "G-9D511367MB",
  apiKey: "AIzaSyCEQgSH4Vfzmbn6IVFOE5fu5DtH0p3pAZ8",

  authDomain: "forum-ovidiu-borze.firebaseapp.com",

  projectId: "forum-ovidiu-borze",

  storageBucket: "forum-ovidiu-borze.appspot.com",

  messagingSenderId: "555287825535",

  appId: "1:555287825535:web:98396be2b79a49d0fcc34a",

  measurementId: "G-71MX5H8V0X"

};

// Initialize Firebase

const app = initializeApp(firebaseConfig);
export const storage = getStorage(app);

export default app;
