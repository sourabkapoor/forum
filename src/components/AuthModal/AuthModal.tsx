import {
  createUserWithEmailAndPassword,
  getAuth,
  signInWithEmailAndPassword,
  updateCurrentUser,
  updateProfile,
} from "firebase/auth";
import React, { useState } from "react";
import CSSModules from "react-css-modules";
import { Link } from "react-router-dom";
import {
  selectSignInModalState,
  selectSignUpModalState,
  toggleSignInModal,
  toggleSignUpModal,
} from "../../features/auth/authSlice";
import { auth, getUser, getUserName, signIn } from "../../firebase";
import { useAppDispatch, useAppSelector } from "../../hooks/hooks";
import Modal from "../Modal/Modal";
import styles from "./AuthModal.module.css";
import exitIcon from "../../assets/exit-icon.svg";
import googleIcon from "../../assets/google-logo.svg";
import { serverTimestamp } from "firebase/firestore";

type InputEvent = React.ChangeEvent<HTMLInputElement>;

type FormEvent = React.FormEvent<HTMLFormElement>;

const AuthModal: React.FC = () => {
  const [email, setEmail] = useState("");
  const [username, setUserName] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");

  const signInModalState = useAppSelector(selectSignInModalState);
  const signUpModalState = useAppSelector(selectSignUpModalState);

  const dispatch = useAppDispatch();

  function handleEmail(event: InputEvent) {
    setEmail(event.target.value);
  }

  function handleUserName(event: InputEvent) {
    setUserName(event.target.value);
  }

  function handlePassword(event: InputEvent) {
    setPassword(event.target.value);
  }

  function handleConfirmPassword(event: InputEvent) {
    setConfirmPassword(event.target.value);
  }

  function handleModalSwitch() {
    dispatch(toggleSignUpModal());
    dispatch(toggleSignInModal());
  }

  function handleModalExit() {
    if (signInModalState === true) {
      dispatch(toggleSignInModal());
    }
    if (signUpModalState === true) {
      dispatch(toggleSignUpModal());
    }
  }

  function onOAuthClick() {
    if (signUpModalState) {
      dispatch(toggleSignUpModal());
    } else if (signInModalState) {
      dispatch(toggleSignInModal());
    }
    signIn();
  }

  async function formSubmit(event: FormEvent) {
    event.preventDefault();

    if (signUpModalState) {
      if (password === confirmPassword) {
        try {
          const userCredential = await createUserWithEmailAndPassword(
            auth,
            email,
            password
          );
          await updateProfile(userCredential.user, {
            displayName: username,
          });
          dispatch(toggleSignUpModal());
        } catch (error) {
          alert(`Sorry, this email has already been used: ${error}`);
        }
      } else {
        alert("Please check your passwords again!");
      }
    } else if (signInModalState) {
      try {
        await signInWithEmailAndPassword(auth, email, password);
        dispatch(toggleSignInModal());
      } catch (error) {
        alert(`Wrong email or password. Please try again: ${error}`);
      }
    }
  }

  return (
    <Modal>
      <div styleName="sign-up-modal">
        <div styleName="sign-up-modal__container">
          <button
            styleName="sign-up-modal__exit"
            onClick={handleModalExit}
            data-testid="exit-button"
          >
            <img
              styleName="sign-up-modal__exit-icon"
              src={exitIcon}
              alt="exit button for modal"
            />
          </button>
          <div styleName="sign-up-modal__image"></div>
          <div styleName="sign-up-modal__content">
            <form styleName="sign-up-modal__form" onSubmit={formSubmit}>
              <h1 styleName="sign-up-modal__title">
                {(signInModalState && "Log In") ||
                  (signUpModalState && "Sign Up")}
              </h1>
              <p styleName="sign-up-modal__description">
                By continuing,{" "}
                {(signInModalState && "you") ||
                  (signUpModalState &&
                    "you are setting up a Forum account and")}{" "}
                agree to our{" "}
                <a
                  styleName="sign-up-modal__link"
                  href="/"
                >
                  User Agreement
                </a>{" "}
                and{" "}
                <a
                  styleName="sign-up-modal__link"
                  href="/"
                >
                  Privacy Policy
                </a>
                .
              </p>
              <button styleName="sign-up-modal__oauth" onClick={onOAuthClick}>
                <img
                  styleName="sign-up-modal__icon"
                  src={googleIcon}
                  alt="google icon"
                />
                <p styleName="sign-up-modal__oauth-text">
                  {" "}
                  Continue with Google
                </p>
              </button>
              <div styleName="sign-up-modal__divider">
                <span styleName="sign-up-modal__divider-line"></span>
                <span styleName="sign-up-modal__divider-text">OR</span>
                <span styleName="sign-up-modal__divider-line"></span>
              </div>
              <input
                type="email"
                styleName="sign-up-modal__input"
                placeholder="EMAIL"
                value={email}
                onChange={handleEmail}
                required
              />
              {signUpModalState && (
                <input
                  styleName="sign-up-modal__input"
                  placeholder="USERNAME"
                  value={username}
                  onChange={handleUserName}
                  required
                />
              )}
              <input
                styleName="sign-up-modal__input"
                type="password"
                placeholder="PASSWORD"
                value={password}
                onChange={handlePassword}
                required
              />
              {signUpModalState && (
                <input
                  styleName="sign-up-modal__input"
                  type="password"
                  placeholder="CONFIRM PASSWORD"
                  required
                  onChange={handleConfirmPassword}
                  value={confirmPassword}
                />
              )}
              <button styleName="sign-up-modal__button" type="submit">
                Continue
              </button>
              <div styleName="sign-up-modal__bottom-text-container">
                {(signInModalState && "New to Forum?") ||
                  (signUpModalState && " Already a Forumer?")}
                <button
                  styleName="sign-up-modal__bottom-text"
                  onClick={handleModalSwitch}
                >
                  {(signInModalState && "SIGN UP") ||
                    (signUpModalState && "LOG IN")}
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </Modal>
  );
};

export default CSSModules(AuthModal, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "log",
});
