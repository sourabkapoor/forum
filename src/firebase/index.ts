import { initializeApp } from "firebase/app";
import {
  getAuth,
  onAuthStateChanged,
  GoogleAuthProvider,
  signInWithPopup,
  signOut,
  Auth,
} from "firebase/auth";
import {
  getFirestore,
  collection,
  addDoc,
  query,
  orderBy,
  limit,
  onSnapshot,
  setDoc,
  updateDoc,
  doc,
  serverTimestamp,
} from "firebase/firestore";
import {
  getStorage,
  ref,
  uploadBytesResumable,
  getDownloadURL,
} from "firebase/storage";
import { getMessaging, getToken, onMessage } from "firebase/messaging";
import { getPerformance } from "firebase/performance";
import { getAnalytics } from "firebase/analytics";

import app from "./firebase-config";
import { authStateObserver } from "../redux/store";

export const auth = getAuth(app);

export const db = getFirestore(app);

export async function signIn() {
  const provider = new GoogleAuthProvider();
  await signInWithPopup(auth, provider);
}

export async function signOutUser() {
  await signOut(auth);
}

export function initFirebaseAuth() {
  onAuthStateChanged(auth, authStateObserver);
}

export function getUserName() {
  return auth.currentUser?.displayName;
}

export function isUserSignedIn() {
  return !!auth.currentUser;
}

export function getUserId() {
  return auth.currentUser?.uid;
}

export function getUser() {
  return auth.currentUser;
}

initFirebaseAuth();

const analytics = getAnalytics(app);

