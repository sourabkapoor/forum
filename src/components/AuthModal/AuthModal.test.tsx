import AuthModal from "./AuthModal";
import {
  getByPlaceholderText,
  queryByPlaceholderText,
  queryByRole,
  queryByTestId,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import StoreProvider from "../../redux/provider";
import * as firebase from "../../firebase";

describe("AuthModal", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <StoreProvider>
        <AuthModal />
      </StoreProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("user can exit modal", async () => {
    const mockSignInModalState = jest.fn().mockReturnValue(true);

    const mockAppDispatch = jest.fn();

    jest.mock("../../hooks/hooks", () => ({
      ...jest.requireActual("../../hooks/hooks"),
      useAppSelector: () => mockSignInModalState,
      useAppDispatch: () => mockAppDispatch,
    }));

    render(
      <StoreProvider>
        <AuthModal />
      </StoreProvider>
    );

    const exitButton = screen.queryByTestId("exit-button");

    await userEvent.click(exitButton as HTMLButtonElement);

    expect(mockAppDispatch).toHaveBeenCalledTimes(1);
  });
  test("user can correctly enter into the input fields", async () => {
    const mockSignInModalState = jest.fn().mockReturnValue(true);

    jest.mock("../../hooks/hooks", () => ({
      ...jest.requireActual("../../hooks/hooks"),
      useAppSelector: () => mockSignInModalState,
    }));

    render(
      <StoreProvider>
        <AuthModal />
      </StoreProvider>
    );

    const emailInput = screen.queryByPlaceholderText(
      "EMAIL"
    ) as HTMLInputElement;
    const userNameInput = screen.queryByPlaceholderText(
      "USERNAME"
    ) as HTMLInputElement;
    const passwordInput = screen.queryByPlaceholderText(
      "PASSWORD"
    ) as HTMLInputElement;
    const confirmPasswordInput = screen.queryByPlaceholderText(
      "CONFIRM PASSWORD"
    ) as HTMLInputElement;

    await userEvent.type(emailInput, "bobjones@gmail.com");
    await userEvent.type(userNameInput, "bobjones123");
    await userEvent.type(passwordInput, "locospollos");
    await userEvent.type(confirmPasswordInput, "locospollos");

    expect(emailInput.textContent).toEqual("bobjones@gmail.com");
    expect(userNameInput.textContent).toEqual("bobjones123");
    expect(passwordInput.textContent).toEqual("locospollos");
    expect(confirmPasswordInput.textContent).toEqual("locospollos");
  });

  test("oauth", async () => {
    render(
      <StoreProvider>
        <AuthModal />
      </StoreProvider>
    );
    const mockSignIn = jest.fn();
    jest.mock("../../firebase", () => ({ signIn: mockSignIn }));

    jest.spyOn(firebase, "signIn");

    const user = userEvent.setup();

    const oAuthButton = screen.queryByRole("button", {
      name: /Continue with Google/i,
    });

    await user.click(oAuthButton!);

    expect(mockSignIn).toBeCalledTimes(1);
  });
  test("email and pasword auth", async () => {
    const user = userEvent.setup();
    const emailInput = screen.queryByPlaceholderText("EMAIL") as HTMLElement;
    const userNameInput = screen.queryByPlaceholderText(
      "USERNAME"
    ) as HTMLInputElement;
    const passwordInput = screen.queryByPlaceholderText(
      "PASSWORD"
    ) as HTMLInputElement;
    const confirmPasswordInput = screen.queryByPlaceholderText(
      "CONFIRM PASSWORD"
    ) as HTMLInputElement;

    await user.type(emailInput, "bobjones@gmail.com");
    await user.type(userNameInput, "bobjones123");
    await user.type(passwordInput, "locospollos");
    await user.type(confirmPasswordInput, "locospollos");
  });
});
