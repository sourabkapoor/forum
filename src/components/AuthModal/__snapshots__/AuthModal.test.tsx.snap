// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModal snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal"
  >
    <div
      class="sign-up-modal"
    >
      <div
        class="sign-up-modal__container"
      >
        <button
          class="sign-up-modal__exit"
          data-testid="exit-button"
        >
          <img
            alt="exit button for modal"
            class="sign-up-modal__exit-icon"
            src="exit-icon.svg"
          />
        </button>
        <div
          class="sign-up-modal__image"
        />
        <div
          class="sign-up-modal__content"
        >
          <form
            class="sign-up-modal__form"
          >
            <h1
              class="sign-up-modal__title"
            />
            <p
              class="sign-up-modal__description"
            >
              By continuing,  agree to our 
              <a
                class="sign-up-modal__link"
                href="https://www.redditinc.com/policies/user-agreement"
              >
                User Agreement
              </a>
               and 
              <a
                class="sign-up-modal__link"
                href="https://www.reddit.com/policies/privacy-policy"
              >
                Privacy Policy
              </a>
              .
            </p>
            <button
              class="sign-up-modal__oauth"
            >
              <img
                alt="google icon"
                class="sign-up-modal__icon"
                src="google-logo.svg"
              />
              <p
                class="sign-up-modal__oauth-text"
              >
                 Continue with Google
              </p>
            </button>
            <div
              class="sign-up-modal__divider"
            >
              <span
                class="sign-up-modal__divider-line"
              />
              <span
                class="sign-up-modal__divider-text"
              >
                OR
              </span>
              <span
                class="sign-up-modal__divider-line"
              />
            </div>
            <input
              class="sign-up-modal__input"
              placeholder="EMAIL"
              required=""
              type="email"
              value=""
            />
            <input
              class="sign-up-modal__input"
              placeholder="PASSWORD"
              required=""
              type="password"
              value=""
            />
            <button
              class="sign-up-modal__button"
              type="submit"
            >
              Continue
            </button>
            <div
              class="sign-up-modal__bottom-text-container"
            >
              <button
                class="sign-up-modal__bottom-text"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
